experimental/learn_descriptors/hloc_helper.py

```diff
@@ -0,0 +1,254 @@
+        if self.path_local_matches_sfm is None:
+            self.path_local_matches_sfm = self.dir_outputs / "matches.h5"
+
+
+class HlocHelper:
+    def __init__(
+        self,
+        config: HlocHelperConfig | None = None,
+        dir_cached_colmap_model: pycolmap.Reconstruction | None = None,
+    ):
+        if config is not None:
+            self.config = config
+        if dir_cached_colmap_model is not None:
+            self.model = pycolmap.Reconstruction(str(dir_cached_colmap_model))
+        else:
+            self.model = None
+
+    # images in image_list should be relative to self.config.dir_images
+    def extract_and_match(self, image_list: list[Path]):
+        extract_features.main(
+            self.config.local_feature_config,
+            self.config.dir_images,
+            self.config.dir_outputs,
+            feature_path=self.config.path_local_features,
+            image_list=image_list,
+        )
+        if not self.config.exhaustive_match:
+            extract_features.main(
+                self.config.global_feature_config,
+                self.config.dir_images,
+                self.config.dir_outputs,
+                feature_path=self.config.path_global_features,
+            )
+            pairs_from_retrieval.main(
+                self.config.path_global_features,
+                self.config.path_sfm_pairs,
+                num_matched=self.config.retrieval_num_matched,
+                query_list=image_list,
+                db_list=image_list,
+            )
+        else:
+            pairs_from_exhaustive.main(
+                self.config.path_sfm_pairs,
+                image_list=image_list,
+            )
+        match_features.main(
+            self.config.matcher_local_config,
+            self.config.path_sfm_pairs,
+            self.config.path_local_features,
+            self.config.dir_outputs,
+            self.config.path_local_matches_sfm,
+        )
+
+    # if image_list is left as None, all images will be imported into colmap_db
+    def reconstruct(self, image_list: list[Path] = None) -> pycolmap.Reconstruction:
+        self.model = reconstruction.main(
+            self.config.dir_sfm,
+            self.config.dir_images,
+            self.config.path_sfm_pairs,
+            self.config.path_local_features,
+            self.config.path_local_matches_sfm,
+            image_list=image_list,
+        )
+        return self.model
+
+    # db_images should be mapped/localized in self.model
+    # both query_images and db_images should be relative paths from self.config.dir_images
+    # (and might have to be children of self self.config.dir_images)
+    def localize_sfm(
+        self,
+        query_images: list[Path],
+        db_images: list[Path],
+        path_localize_pairs: Path,
+        path_results: Path,
+        query_images_intrinsics: Optional[List[List[str]]] = None,
+    ):
+        assert self.model is not None
```

Comment:
the `db_images` and the config can be used to identify a model. If the model with the given config and `db_images` hasn't been created, you can assert the desired model hasn't been reconstructed.